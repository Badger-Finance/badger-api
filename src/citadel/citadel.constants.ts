import { Protocol } from '@badger-dao/sdk';

export const CITADEL_SUBGRAPH_URL = 'https://api.thegraph.com/subgraphs/name/axejintao/citadog';
export const CITADEL_KNIGHTS_ROUND_SUBGRAPH_URL =
  'https://api.thegraph.com/subgraphs/name/namelsking/citadel-knighting-rounds';

export const CITADEL_KNIGHTS = [
  Protocol.Terra,
  Protocol.Frax,
  Protocol.Alchemix,
  Protocol.Convex,
  Protocol.Redacted,
  Protocol.JonesDao,
  Protocol.Tokemak,
  Protocol.Tribe,
  Protocol.Ren,
];
