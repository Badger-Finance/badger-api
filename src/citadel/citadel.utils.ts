import { ethers } from 'ethers';
import { getDataMapper } from '../aws/dynamodb.utils';
import { KeyedDataBlob } from '../aws/models/keyed-data-blob.model';
import { CitadelData, CTIADEL_DATA } from './destructors/citadel-data.destructor';
import { Nullable } from '../utils/types.utils';
import { CitadelRewardsSnapshot } from '../aws/models/citadel-rewards-snapshot';
import BadgerSDK from '@badger-dao/sdk';
import { RewardEventType, RewardEventTypeEnum } from '@badger-dao/sdk/lib/citadel/enums/reward-event-type.enum';
import { ListRewardsEvent } from '@badger-dao/sdk/lib/citadel/interfaces/list-rewards-event.interface';
import { CitadelRewardType } from '@badger-dao/sdk/lib/api/enums/citadel-reward-type.enum';
import { CitadelRewardsAprBlob } from './interfaces/citadel-rewards-apr-blob.interface';

export async function queryCitadelData(): Promise<CitadelData> {
  const mapper = getDataMapper();

  let blob;
  for await (const item of mapper.query(KeyedDataBlob, { id: CTIADEL_DATA }, { limit: 1 })) {
    blob = item;
  }
  if (!blob) {
    throw new Error('No stored citadel data!');
  }

  return new CitadelData(blob.data);
}

export async function getRewardsOnchain(
  sdk: BadgerSDK,
  type: RewardEventType,
  fromLastBlock = false,
): Promise<ListRewardsEvent[]> {
  let chainRewards: ListRewardsEvent[] = [];

  let lastRewardBlock: Nullable<number>;

  if (fromLastBlock) lastRewardBlock = Number((await getLastRewardByType(type))?.block) + 1;

  try {
    const getOpts: { filter: RewardEventType; startBlock?: number } = { filter: type };
    if (lastRewardBlock) getOpts.startBlock = lastRewardBlock;

    chainRewards = await sdk.citadel.listRewards(getOpts);
  } catch (err) {
    throw new Error(`Failed to get citadel listRewards ${err}`);
  }

  return chainRewards;
}

export async function getLastRewardByType(type: string): Promise<Nullable<CitadelRewardsSnapshot>> {
  const mapper = getDataMapper();

  const query = mapper.query(
    CitadelRewardsSnapshot,
    { payType: type },
    { limit: 1, scanIndexForward: false, indexName: 'IndexCitadelRewardsDataPayTypeBlock' },
  );

  let lastReward = null;

  try {
    for await (const reward of query) {
      lastReward = reward;
    }
  } catch (e) {
    console.error(`Failed to get citadel reward from ddb for type: ${type}; ${e}`);
  }

  return lastReward;
}

export async function getRewardsAprForDataBlob(): Promise<CitadelRewardsAprBlob> {
  const mapper = getDataMapper();

  const summaryCountTpl = {
    apr: 0,
    count: 0,
  };
  const summary = {
    overall: { ...summaryCountTpl },
    [CitadelRewardType.Citadel]: { ...summaryCountTpl },
    [CitadelRewardType.Funding]: { ...summaryCountTpl },
    [CitadelRewardType.Tokens]: { ...summaryCountTpl },
    [CitadelRewardType.Yield]: { ...summaryCountTpl },
  };

  const query = mapper.query(
    CitadelRewardsSnapshot,
    { payType: RewardEventTypeEnum.ADDED },
    {
      indexName: 'IndexCitadelRewardsDataPayType',
      filter: {
        type: 'GreaterThanOrEqualTo',
        object: Date.now(),
        subject: 'finishTime',
      },
    },
  );

  try {
    for await (const reward of query) {
      summary.overall.apr += reward.apr || 0;
      summary.overall.count += 1;

      const rewardTypeKey = getRewardsEventTypeMapped(<string>reward.dataType);

      summary[rewardTypeKey].apr += reward.apr || 0;
      summary[rewardTypeKey].count += 1;
    }
  } catch (e) {
    console.error(`Failed to get citadel reward from ddb`);
  }

  return Object.keys(summary).reduce((acc, val) => {
    const assertKey = <keyof CitadelRewardsAprBlob>val;
    acc[assertKey] = summary[assertKey].apr / summary[assertKey].count;
    return acc;
  }, <CitadelRewardsAprBlob>{});
}

export function dataTypeRawKeyToKeccak(key: string): string {
  let hashedKey: string;

  try {
    hashedKey = ethers.utils.keccak256(ethers.utils.formatBytes32String(key));
  } catch (e) {
    console.warn(`Failed to hash raw key: ${key}. ${e}`);
    hashedKey = '';
  }

  return hashedKey;
}

export function getRewardsEventTypeMapped(hashKey: string): CitadelRewardType {
  return (
    {
      // Note! This is still not actual keys, need to figure out, what value
      // is hashed in the event
      [dataTypeRawKeyToKeccak('Citadel')]: CitadelRewardType.Citadel,
      [dataTypeRawKeyToKeccak('Funding')]: CitadelRewardType.Funding,
      [dataTypeRawKeyToKeccak('Yield')]: CitadelRewardType.Yield,
      [dataTypeRawKeyToKeccak('Tokens')]: CitadelRewardType.Tokens,
    }[hashKey] || CitadelRewardType.Tokens
  );
}
