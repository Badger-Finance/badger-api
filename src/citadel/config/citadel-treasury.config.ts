import { Network } from '@badger-dao/sdk';
import { SUPPORTED_CHAINS } from '../../chains/chain';

const citadelChains = new Set([Network.Ethereum]);

const chains = SUPPORTED_CHAINS.filter((c) => citadelChains.has(c.network));
const tokens = new Set(chains.flatMap((c) => Object.keys(c.tokens)));
const vaults = new Set(chains.flatMap((c) => c.vaults.map((v) => v.vaultToken)));
vaults.forEach((v) => tokens.delete(v));

export const TRACKED_TOKENS = [...tokens];
export const TRACKED_VAULTS = [...vaults];

export const CITADEL_TREASURY_ADDRESS = '0x38724146C8dc1Aa49c3395091cf86B789c37F52c';
