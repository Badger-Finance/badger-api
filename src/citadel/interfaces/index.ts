export { CitadelTreasurySummary } from './citadel-treasury-summary.interface';
