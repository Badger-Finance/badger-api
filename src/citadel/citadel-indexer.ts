import { Erc20__factory, formatBalance, Protocol } from '@badger-dao/sdk';
import { getCachedAccount } from '../accounts/accounts.utils';
import { Ethereum } from '../chains/config/eth.config';
import { getPrice } from '../prices/prices.utils';
import { VaultsService } from '../vaults/vaults.service';
import { getVaultDefinition } from '../vaults/vaults.utils';
import { CITADEL_TREASURY_ADDRESS, TRACKED_TOKENS, TRACKED_VAULTS } from './config/citadel-treasury.config';
import { TreasuryPosition } from '../treasury/interfaces/treasy-position.interface';
import { TreasurySummary } from '../treasury/interfaces/treasury-summary.interface';
import { getDataMapper } from '../aws/dynamodb.utils';
import { TreasurySummarySnapshot } from '../aws/models/treasury-summary-snapshot.model';
import { TOKEN_PRICE_DATA, TREASURY_HISTORIC_DATA, TREASURY_SNAPSHOT_DATA } from '../config/constants';

export async function snapshotTreasury() {
  const chain = new Ethereum();
  const sdk = await chain.getSdk();

  const lookupTokens = TRACKED_TOKENS.concat(TRACKED_VAULTS);
  const tokenInformation = await sdk.tokens.loadTokens(lookupTokens);

  const vaultPositions = new Set(TRACKED_VAULTS);
  const treasuryPositions: TreasuryPosition[] = await Promise.all(
    Object.entries(tokenInformation).map(async (e) => {
      const [address, token] = e;
      const isFarming = vaultPositions.has(address);

      const contract = Erc20__factory.connect(address, sdk.provider);
      const treasuryBalance = await contract.balanceOf(CITADEL_TREASURY_ADDRESS);

      const { price } = await getPrice(address);
      const balance = formatBalance(treasuryBalance, token.decimals);
      const value = price * balance;

      if (!isFarming) {
        return {
          ...token,
          balance,
          value,
          apr: 0,
          protocol: 'none',
        };
      }

      const { apr, minApr, maxApr } = await VaultsService.loadVault(chain, getVaultDefinition(chain, address));
      const { boost } = await getCachedAccount(chain, CITADEL_TREASURY_ADDRESS);
      const percentile = boost / 2000;

      let treasuryApr;
      if (minApr && maxApr) {
        treasuryApr = minApr + (maxApr - minApr) * percentile;
      } else {
        treasuryApr = apr;
      }

      return {
        ...token,
        balance,
        value,
        apr: treasuryApr,
        protocol: Protocol.Badger,
      };
    }),
  );

  const positions = treasuryPositions.filter((p) => p.value > 0);
  const value = positions.reduce((total, position) => (total += position.value), 0);

  let tvlApr = 0;
  positions.forEach((p) => {
    if (p.apr > 0) {
      tvlApr += p.apr * p.value;
    }
  });

  const treasuryYield = value === 0 ? 0 : tvlApr / value;

  const treasurySummary: TreasurySummary = {
    address: CITADEL_TREASURY_ADDRESS,
    value,
    yield: treasuryYield,
    positions,
  };

  const mapper = getDataMapper();

  console.log({ TREASURY_SNAPSHOT_DATA, TREASURY_HISTORIC_DATA, TOKEN_PRICE_DATA });
  // capture a current view of the treasury
  try {
    const currentSnapshot = Object.assign(new TreasurySummarySnapshot(), treasurySummary);
    console.log(currentSnapshot);
    await mapper.put(currentSnapshot);
  } catch (err) {
    console.error({ message: 'Unable to save Ctiadel treasury snapshot', err });
  }
}
