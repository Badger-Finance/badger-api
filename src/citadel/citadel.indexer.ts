import { formatBalance } from '@badger-dao/sdk';
import { Ethereum } from '../chains/config/eth.config';
import { TRACKED_TOKENS, TRACKED_VAULTS } from './config/citadel-treasury.config';
import { TreasuryPosition } from './interfaces/treasy-position.interface';

export async function snapshotTreasury() {
  const chain = new Ethereum();
  const sdk = await chain.getSdk();

  const lookupTokens = TRACKED_TOKENS.concat(TRACKED_VAULTS);
  const tokenBalances = await sdk.tokens.loadBalances(lookupTokens);
  const tokenInformation = await sdk.tokens.loadTokens(lookupTokens);

  const vaultPositions = new Set(TRACKED_VAULTS);
  const treasuryPositions: TreasuryPosition[] = Object.entries(tokenBalances).map((e) => {
    const [token, balance] = e;
    const isFarming = vaultPositions.has(token);
    const tokenInfo = tokenInformation[token];
    if (!isFarming) {
      return {
        token: tokenInfo,
        amount: formatBalance(balance, tokenInfo.decimals),
        value: 0,
        apr: 0,
      };
    }
    return {
      token: tokenInfo,
      amount: formatBalance(balance, tokenInfo.decimals),
      value: 0,
      apr: 0,
    };
  });
}
