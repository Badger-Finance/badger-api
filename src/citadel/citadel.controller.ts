import { CitadelTreasurySummary } from '@badger-dao/sdk/lib/api/interfaces/citadel-treasury-summary.interface';
import { Controller, Get, Inject, QueryParams, UseCache } from '@tsed/common';
import { ContentType, Summary, Returns, Description } from '@tsed/schema';
import { HistoricTreasurySummarySnapshot } from '../aws/models/historic-treasury-summary-snapshot.model';
import { ChartTimeFrame } from '../charts/enums/chart-timeframe.enum';
import { TreasuryService } from '../treasury/treasury.service';
import { CitadelService } from './citadel.service';
import { CITADEL_TREASURY_ADDRESS } from './config/citadel-treasury.config';
import { CitadelTreasurySummaryModel } from './interfaces/citadel-treasury-summary-model.interface';
import { RewardFilter } from '@badger-dao/sdk/lib/citadel/enums/reward-filter.enum';
import { CitadelRewardEventModel } from './interfaces/citadel-reward-event-model.interface';
import { CitadelRewardEvent } from './interfaces/citadel-reward-event.interface';

@Controller('/')
export class CitadelController {
  @Inject()
  citadelService!: CitadelService;

  @Inject()
  treasuryService!: TreasuryService;

  @UseCache()
  @Get('/treasury')
  @ContentType('json')
  @Summary('Citadel DAO Treasury Allocation and Metrics')
  @Description('Returns specific treasury metrics, positions, and yield')
  @Returns(200, CitadelTreasurySummaryModel)
  async loadCitadelTreasury(): Promise<CitadelTreasurySummary> {
    return this.citadelService.loadTreasurySummary();
  }

  @Get('/rewards')
  @ContentType('json')
  @Summary('Added/paid rewards list')
  @Description('List of reward events')
  @Returns(200, Array).Of(CitadelRewardEventModel)
  @Returns(400).Description('Token should be specified')
  @Returns(400).Description('User or token param is missing')
  async loadRewardsList(
    @QueryParams('token') token: string,
    @QueryParams('user') user?: string,
    @QueryParams('filter') filter?: RewardFilter,
  ): Promise<CitadelRewardEvent[]> {
    return this.citadelService.getListRewards(token, user, filter);
  }

  @UseCache()
  @Get('/history')
  @ContentType('json')
  async loadCitadelTreasuryCharts(
    @QueryParams('timeframe') timeframe = ChartTimeFrame.Day,
  ): Promise<HistoricTreasurySummarySnapshot[]> {
    return this.treasuryService.loadTreasuryChartData(CITADEL_TREASURY_ADDRESS, timeframe);
  }
}
