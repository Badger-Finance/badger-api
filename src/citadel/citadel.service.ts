import { Erc20__factory, formatBalance } from '@badger-dao/sdk';
import { CitadelTreasurySummary } from '@badger-dao/sdk/lib/api/interfaces/citadel-treasury-summary.interface';
import { Service } from '@tsed/di';
import { TOKENS } from '../config/tokens.config';
import { getPrice } from '../prices/prices.utils';
import { queryTreasurySummary } from '../treasury/treasury.utils';
import { queryCitadelData } from './citadel.utils';
import { CITADEL_TREASURY_ADDRESS } from './config/citadel-treasury.config';
import { Chain } from '../chains/config/chain.config';
import { RewardFilter } from '@badger-dao/sdk/lib/citadel/enums/reward-filter.enum';
import { BadRequest } from '@tsed/exceptions';
import { ListRewardsEvent } from '@badger-dao/sdk/lib/citadel/interfaces/list-rewards-event.interface';
import { CitadelRewardEvent } from './interfaces/citadel-reward-event.interface';

@Service()
export class CitadelService {
  async loadTreasurySummary(): Promise<CitadelTreasurySummary> {
    const [baseTreasurySummary, citadelData] = await Promise.all([
      queryTreasurySummary(CITADEL_TREASURY_ADDRESS),
      queryCitadelData(),
    ]);

    const { price } = await getPrice(TOKENS.WBTC);
    const {
      marketCapToTreasuryRatio,
      valuePaid,
      fundingBps,
      lockingBps,
      stakingBps,
      marketCap,
      supply,
      staked,
      stakedPercent,
    } = citadelData;

    const valueBtc = baseTreasurySummary.value / price;
    const valuePaidBtc = valuePaid / price;

    return {
      ...baseTreasurySummary,
      valueBtc,
      valuePaid,
      valuePaidBtc,
      marketCapToTreasuryRatio,
      fundingBps,
      stakingBps,
      lockingBps,
      marketCap,
      supply,
      staked,
      stakedPercent,
    };
  }

  async getListRewards(token: string, user?: string, filter?: RewardFilter): Promise<CitadelRewardEvent[]> {
    const chain = Chain.getChain();
    const sdk = await chain.getSdk();

    let chainRewards: ListRewardsEvent[] = [];

    try {
      chainRewards = await sdk.citadel.listRewards({ user, token, filter });
    } catch (err) {
      throw new BadRequest(`${err}`);
    }

    return Promise.all(
      chainRewards.map(async (event) => {
        const rewardEventResp: CitadelRewardEvent = {
          token: event.token,
          block: event.block,
          amount: 0,
          type: filter ?? RewardFilter.ADDED,
        };

        const tokenContract = Erc20__factory.connect(event.token, sdk.provider);
        const tokenDecimals = await tokenContract.decimals();

        rewardEventResp.amount = formatBalance(event.reward, tokenDecimals);

        if (event.user) rewardEventResp.user = event.user;

        return rewardEventResp;
      }),
    );
  }
}
