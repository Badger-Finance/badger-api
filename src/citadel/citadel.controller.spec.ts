import { BadRequest } from '@tsed/exceptions';

import { CitadelService } from './citadel.service';
import SuperTest from 'supertest';
import { PlatformTest } from '@tsed/common';
import { Server } from '../Server';

import citadelTreasuryMock from '@badger-dao/sdk-mocks/generated/ethereum/api/loadCitadelTreasury.json';
import { TOKENS } from '../config/tokens.config';
import { RewardFilter } from '@badger-dao/sdk/lib/citadel/enums/reward-filter.enum';
import { citadelRewardEventsMock } from './mocks/citadel-reward-events.mock';
import { CitadelTreasurySummary } from '@badger-dao/sdk';

describe('CitadelController', () => {
  let request: SuperTest.SuperTest<SuperTest.Test>;

  beforeAll(PlatformTest.bootstrap(Server));
  beforeAll(async () => {
    request = SuperTest(PlatformTest.callback());
  });
  afterAll(PlatformTest.reset);

  describe('GET /citadel/v1/treasury', () => {
    it('it returns the current citadel treasury summary', async (done: jest.DoneCallback) => {
      // TODO: remove to simply the mock once the sdk mocks are updated
      const expected: CitadelTreasurySummary = {
        ...citadelTreasuryMock,
        staked: 0,
        stakedPercent: 0,
        supply: 10,
        marketCap: 100,
      };
      jest.spyOn(CitadelService.prototype, 'loadTreasurySummary').mockImplementation(async () => expected);
      const { body } = await request.get('/citadel/v1/treasury').expect(200);
      expect(body).toMatchObject(expected);
      done();
    });
  });

  describe('GET /citadel/v1/rewards', () => {
    const TEST_TOKEN = TOKENS.CTDL;
    const TEST_USER = '0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28';

    it('should return rewards for user', async (done: jest.DoneCallback) => {
      // TODO: Replace with Citadel service mocks from sdk, once rewards are available
      jest.spyOn(CitadelService.prototype, 'getListRewards').mockImplementation(async () => citadelRewardEventsMock);

      const { body } = await request
        .get(`/citadel/v1/rewards?filter=${RewardFilter.PAID}&token=${TEST_TOKEN}&user=${TEST_USER}`)
        .expect(200);
      expect(body).toMatchSnapshot();
      done();
    });

    it('should fail without token addr param', async (done: jest.DoneCallback) => {
      const { body } = await request.get('/citadel/v1/rewards').expect(BadRequest.STATUS);
      expect(body).toMatchSnapshot();
      done();
    });

    it('should fail without user addr param for paid filter', async (done: jest.DoneCallback) => {
      const { body } = await request
        .get(`/citadel/v1/rewards?filter=${RewardFilter.PAID}&token=${TEST_TOKEN}`)
        .expect(BadRequest.STATUS);
      expect(body).toMatchSnapshot();
      done();
    });
  });
});
