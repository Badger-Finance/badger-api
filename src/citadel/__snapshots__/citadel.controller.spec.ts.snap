// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitadelController GET /citadel/v1/rewards should fail without token addr param 1`] = `
Object {
  "message": "Error: processing response error (body=\\"{\\\\\\"jsonrpc\\\\\\":\\\\\\"2.0\\\\\\",\\\\\\"error\\\\\\":{\\\\\\"code\\\\\\":-32600,\\\\\\"message\\\\\\":\\\\\\"Invalid Request. Requested data is older than 128 blocks.\\\\\\"},\\\\\\"id\\\\\\":51}\\", error={\\"code\\":-32600}, requestBody=\\"{\\\\\\"method\\\\\\":\\\\\\"eth_getLogs\\\\\\",\\\\\\"params\\\\\\":[{\\\\\\"fromBlock\\\\\\":\\\\\\"0xdfb49f\\\\\\",\\\\\\"toBlock\\\\\\":\\\\\\"0xdfe837\\\\\\",\\\\\\"address\\\\\\":\\\\\\"0xb1c38253ad6ab3e2a2d53a094692fcf1321b12d4\\\\\\",\\\\\\"topics\\\\\\":[\\\\\\"0xa3e98ad001ab3597c5c4a0afe6de30207ef7d124a67337acefc3d5d5d905242f\\\\\\"]}],\\\\\\"id\\\\\\":51,\\\\\\"jsonrpc\\\\\\":\\\\\\"2.0\\\\\\"}\\", requestMethod=\\"POST\\", url=\\"https://cloudflare-eth.com\\", code=SERVER_ERROR, version=web/5.6.0)",
}
`;

exports[`CitadelController GET /citadel/v1/rewards should fail without user addr param for paid filter 1`] = `
Object {
  "message": "Error: processing response error (body=\\"{\\\\\\"jsonrpc\\\\\\":\\\\\\"2.0\\\\\\",\\\\\\"error\\\\\\":{\\\\\\"code\\\\\\":-32600,\\\\\\"message\\\\\\":\\\\\\"Invalid Request. Requested data is older than 128 blocks.\\\\\\"},\\\\\\"id\\\\\\":55}\\", error={\\"code\\":-32600}, requestBody=\\"{\\\\\\"method\\\\\\":\\\\\\"eth_getLogs\\\\\\",\\\\\\"params\\\\\\":[{\\\\\\"fromBlock\\\\\\":\\\\\\"0xdfb49f\\\\\\",\\\\\\"toBlock\\\\\\":\\\\\\"0xdfe837\\\\\\",\\\\\\"address\\\\\\":\\\\\\"0xb1c38253ad6ab3e2a2d53a094692fcf1321b12d4\\\\\\",\\\\\\"topics\\\\\\":[\\\\\\"0x540798df468d7b23d11f156fdb954cb19ad414d150722a7b6d55ba369dea792e\\\\\\",null,\\\\\\"0x000000000000000000000000af0b1fdf9c6bfec7b3512f207553c0ba00d7f1a2\\\\\\"]}],\\\\\\"id\\\\\\":55,\\\\\\"jsonrpc\\\\\\":\\\\\\"2.0\\\\\\"}\\", requestMethod=\\"POST\\", url=\\"https://cloudflare-eth.com\\", code=SERVER_ERROR, version=web/5.6.0)",
}
`;

exports[`CitadelController GET /citadel/v1/rewards should return rewards for user 1`] = `
Array [
  Object {
    "amount": 10.234,
    "block": 10,
    "token": "0xaF0b1FDf9c6BfeC7b3512F207553c0BA00D7f1A2",
    "type": "added",
    "user": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
  },
  Object {
    "amount": 3,
    "block": 11,
    "token": "0xaF0b1FDf9c6BfeC7b3512F207553c0BA00D7f1A2",
    "type": "added",
    "user": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
  },
  Object {
    "amount": 0,
    "block": 12,
    "token": "0xaF0b1FDf9c6BfeC7b3512F207553c0BA00D7f1A2",
    "type": "added",
    "user": "0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28",
  },
]
`;
